R/Quosure.R

```diff
@@ -0,0 +1,307 @@
+  standardGeneric("eval_code")
+})
+
+#' @rdname eval_code
+#' @export
+setMethod(
+  "eval_code", signature("Quosure", "character"),
+  function(object, code, name = "code") {
+    # combine code vector (and make names unique)
+    nm <- make.unique(c(names(object@code), name))
+    object@code <- setNames(c(object@code, code), nm)
+
+    # need to copy the objects from old env to new env
+    # to avoid "chunks" sharing the same environment
+    object@env <- .copy_env(object@env)
+    eval(parse(text = code), envir = object@env)
+    lockEnvironment(object@env)
+    object
+  }
+)
+
+#' @rdname eval_code
+#' @export
+setMethod(
+  "eval_code", signature("Quosure", "language"),
+  function(object, code, name = "code") {
+    code_char <- as.expression(code)
+    eval_code(object, code_char, name = name)
+  }
+)
+
+#' @rdname eval_code
+#' @export
+setMethod(
+  "eval_code", signature("Quosure", "ANY"),
+  function(object, code, name = "code") {
+    code_expr <- substitute(code)
+    eval_code(object, code_expr, name = name)
+  }
+)
+
+
+#' @rdname eval_code
+#' @export
+setMethod(
+  "eval_code", signature("Quosure", "expression"),
+  function(object, code, name = "code") {
+    code_char <- as.character(code)
+    eval_code(object, code_char, name = name)
+  }
+)
+
+#' Get object from the `Quosure` environment
+#'
+#' Get object from the `Quosure` environment.
+#' @param object (`Quosure`)
+#' @param var (`character(1)`) name of the variable to pull from the environment.
+#' @name get_var
+#' @export
+setGeneric("get_var", function(object, var) {
+  standardGeneric("get_var")
+})
+
+
+#' @rdname get_var
+#' @export
+setMethod("get_var", signature("Quosure", "character"), function(object, var) {
+  get(var, envir = object@env)
+})
+
+#' @param x (`Quosure`)
+#' @param i (`character`) name of the binding in environment (name of the variable)
+#' @param j not used
+#' @param ... not used
+#' @rdname get_var
+#' @export
+setMethod("[[", c("Quosure", "ANY", "missing"), function(x, i, j, ...) {
+  get_var(x, i)
+})
+
+
+#' Get the code from the `Quosure`
+#'
+#' @param object (`Quosure`)
+setMethod("get_code", signature("Quosure"), function(object) {
+  object@code
+})
+
+#' Join two `Quosure` objects
+#'
+#' Combine two `Quosure` object by merging their environments and the code.
+#' Not every `Quosure` objects can be combined:
+#' - if their environments contains objects of the same name but not identical
+#' - if `object2` has unique code element placed before common element. This means that `object2`
+#' in the environment of the `object2` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases todo
+#' @param object (`Quosure`)
+#' @param object2 (`Quosure`)
+#' @export
+setGeneric("join", function(object, object2, overwrite = FALSE) {
+  standardGeneric("join")
+})
+
+setMethod("join", signature("Quosure", "Quosure"), function(object, object2) {
```

Comment:
I can see that you do more of a _combine_ in which you treat both quosures the same. Do we need a _join_ instead? As a _join_ I mean merge one into the other (vide: joining the tables - right table into the left one). _Join_ would be much more simpler (i.e. `parent.env(right) <- left`) and you don't have to track all the edge cases.

Conceptually, what would be the output of the following (pseudo-code):
```
q1 <- new_quosure("x <- 1", new.env()$x <- 1)
q2 <- new_quosure("x <- 2", new.env()$x <- 2) 
q <- join(q1, q2)
```

You have hard stop. Don't we want `q@env$x == 2; q@code == "x<-1; x<-2"` instead?

Please think of the class role. It needs to track the environment together with the code generating it. The above scenario is perfectly fine workflow thus I would expect it not to fail.

Reply:
I think you are right. If we consider two quosures to be reproducible (code ~ env). Then _q1 ∪ q2_ is a _code1 ∪ code2_ and _env1 ∪ env2_ (on the overlap object from env2 overwrites object from env1).

Reply:
I don't understand the comment with combine, collapse, join. Do you think join is a wrong name?

Reply:
> I don't understand the comment with combine, collapse, join. Do you think join is a wrong name?

For _join_ the argument order matters as you put the _right_ one into the _left_ one - therefore the _right_ one has higher importance that the _left_ one as it might overwrite it.
For combine/collapse you tend to treat both _left_ and _right_ the same way and try to raise an error on conflicts.

Name should reflect the behaviour. Looking at the current function body it's more a combine/collapse than join. But I feel we agree to have join behaviour instead therefore the name could stay as it is now.

Reply:
I think order matters as `y` can overwrite `x`. The most appropriate name in my opinion for them is `union` but probably we shouldn't register a S4 method union as there is already S3. I think `join` also makes sense because we don't combine only but also resolve conflicts between duplicated code and objects.